src/Microsoft.DocAsCode.Dfm/MarkdownValidators/MarkdownValidatorBuilder.cs

```diff
@@ -15,75 +16,202 @@ namespace Microsoft.DocAsCode.Dfm.MarkdownValidators
 
     public class MarkdownValidatorBuilder
     {
+        public const string DefaultValidatorName = "default";
         public const string MarkdownValidatePhaseName = "Markdown style";
 
-        private static readonly Regex OpeningTag = new Regex(@"^\<(\w+)((?:""[^""]*""|'[^']*'|[^'"">])*?)\>$", RegexOptions.Compiled);
-        private static readonly Regex ClosingTag = new Regex(@"^\</(\w+)((?:""[^""]*""|'[^']*'|[^'"">])*?)\>$", RegexOptions.Compiled);
-
         public CompositionHost CompositionHost { get; }
 
-        public ImmutableList<MarkdownTagValidationRule> TagValidators { get; set; }
-
-        public ImmutableList<string> ValidatorContracts { get; set; }
+        private readonly List<RuleWithId<MarkdownTagValidationRule>> _tagValidators =
+            new List<RuleWithId<MarkdownTagValidationRule>>();
+        private readonly List<RuleWithId<MarkdownValidationRule>> _validators =
+            new List<RuleWithId<MarkdownValidationRule>>();
+        private readonly List<MarkdownValidationSetting> _settings =
+            new List<MarkdownValidationSetting>();
+        private readonly Dictionary<string, MarkdownValidationRule> _globalValidators =
+            new Dictionary<string, MarkdownValidationRule>();
 
         public MarkdownValidatorBuilder(CompositionHost host)
         {
+            if (host == null)
+            {
+                throw new ArgumentNullException(nameof(host));
+            }
             CompositionHost = host;
-            TagValidators = ImmutableList<MarkdownTagValidationRule>.Empty;
-            ValidatorContracts = ImmutableList<string>.Empty;
         }
 
-        public void AddTagValidators(params MarkdownTagValidationRule[] validators)
+        public void AddTagValidators(string category, Dictionary<string, MarkdownTagValidationRule> validators)
+        {
+            if (validators == null)
+            {
+                return;
+            }
+            foreach (var pair in validators)
+            {
+                _tagValidators.Add(new RuleWithId<MarkdownTagValidationRule>
+                {
+                    Category = category,
+                    Id = pair.Key,
```

Comment:
just use id in category, because settings always require category.

---
In reply to: [73126663](https://github.com/dotnet/docfx/pull/545#discussion_r73126663) [](ancestors = 73126663)